demos/twa-post-message/src/main/java/com/google/androidbrowserhelper/demos/twapostmessage/MainActivity.java

```diff
@@ -1,122 +1,164 @@
-    Log.d(TAG, "Binding to $packageName.");
-    CustomTabsClient.bindCustomTabsService(this, packageName, new CustomTabsServiceConnection() {
-      @Override
-      public void onCustomTabsServiceConnected(@NonNull ComponentName name,
-          @NonNull CustomTabsClient client) {
-        mClient = client;
-
-        // Note: validateRelationship requires warmup to have been called.
-        client.warmup(0L);
-
-        mSession = mClient.newSession(customTabsCallback);
-        registerBroadcastReceiver();
-      }
-
-      @Override
-      public void onServiceDisconnected(ComponentName componentName) {
-        mClient = null;
-
-      }
-    });
-  }
-
-  private void registerBroadcastReceiver() {
-    IntentFilter intentFilter = new IntentFilter();
-    intentFilter.addAction(PostMessageBroadcastReceiver.POST_MESSAGE_ACTION);
-    registerReceiver(new PostMessageBroadcastReceiver(mSession), intentFilter);
-  }
+import androidx.browser.trusted.TrustedWebActivityIntentBuilder;
+import androidx.core.content.ContextCompat;
+
+/**
+ * A demo to showcase how to use the postMessage in CCT/TWA. Please note that the initialization has
+ * to be from the app, the steps will be as following:
+ *
+ * 1. Bind CustomTabsService and wait for the CustomTabsClient to be ready. 2. Use this Client to
+ * warmup and to create a new CustomTabsSession. 3. Listen for navigation events from
+ * CustomTabsCallback#onNavigationEvent. 4. Upon receiving NAVIGATION_FINISHED, we request a new
+ * PostMessageChannel from CustomTabsSession#requestPostMessageChannel. 5. When the channel is ready
+ * we can initialize the communication by posting the first message using
+ * CustomTabsSession#postMessage.
+ *
+ * Please note that requesting post message channel validates the relationship between your origin
+ * and the application using CustomTabsSession#validateRelationship, with the relation as
+ * Relation#RELATION_USE_AS_ORIGIN, please read the methods doc for more details.
+ *
+ * Validation with the origin doesn't necessarily mean that communication is exclusive to this
+ * origin, but the DAL validation is required to provide MessageEvent.origin field.
+ */
+public class MainActivity extends AppCompatActivity {
+
+    private CustomTabsClient mClient;
+    private CustomTabsSession mSession;
+    private Uri URL = Uri.parse("https://peconn.github.io/starters");
+
+    // This can be the app's package name.
+    private Uri SOURCE_ORIGIN = Uri.parse("my-app-origin-uri");
+    private Uri TARGET_ORIGIN = Uri.parse("https://peconn.github.io");
+    private boolean mValidated = false;
+
+    private final String TAG = "PostMessageDemo";
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_main);
+        // No need to ask for permission as the compileSDK is 31.
+        MessageNotificationHandler.createNotificationChannelIfNeeded(this);
```

Comment:
Could you add a note to the `build.gradle` saying we'll need to change this if we change the compileSdk?